File: bench/perf-budget.ts
import { existsSync } from 'node:fs';

const WASM_PATH = './dist/wasm/bb_wasm.js';
const SNAPSHOT_PATH = './dist/data/snapshot.ubx';

interface WasmModule {
  init(data: Uint8Array): void;
  is_initialized(): boolean;
  should_block(url: string, requestType: string, initiator: string | undefined): boolean;
  get_snapshot_info(): { size: number; initialized: boolean };
}

interface Budget {
  name: string;
  limit: number;
  unit: string;
  actual: number;
  passed: boolean;
}

const BUDGETS = {
  coldStartMs: 500,
  wasmPeakMb: 50,
  matchP99Us: 1000,
  snapshotSizeMb: 30,
};

function assertFileExists(path: string, hint: string): void {
  if (!existsSync(path)) {
    throw new Error(`${hint} not found at ${path}`);
  }
}

async function loadWasm(): Promise<WasmModule> {
  const wasmJsPath = Bun.resolveSync(WASM_PATH, process.cwd());
  const wasmBinaryPath = wasmJsPath.replace('.js', '_bg.wasm');

  assertFileExists(wasmJsPath, 'WASM JS module');
  assertFileExists(wasmBinaryPath, 'WASM binary');

  const jsModule = await import(wasmJsPath);
  const wasmBytes = await Bun.file(wasmBinaryPath).arrayBuffer();

  await jsModule.default({ module_or_path: wasmBytes });

  return jsModule as unknown as WasmModule;
}

async function loadSnapshot(): Promise<Uint8Array> {
  assertFileExists(SNAPSHOT_PATH, 'Snapshot file');
  const file = Bun.file(SNAPSHOT_PATH);
  const buffer = await file.arrayBuffer();
  return new Uint8Array(buffer);
}

function measureColdStart(wasm: WasmModule, snapshot: Uint8Array): number {
  const start = performance.now();
  wasm.init(snapshot);
  return performance.now() - start;
}

function measureMatchLatency(wasm: WasmModule, iterations: number): number[] {
  const testUrls = [
    { url: 'https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js', type: 'script', initiator: 'https://example.com' },
    { url: 'https://www.google-analytics.com/analytics.js', type: 'script', initiator: 'https://example.com' },
    { url: 'https://example.com/style.css', type: 'stylesheet', initiator: 'https://example.com' },
    { url: 'https://cdn.example.com/image.png', type: 'image', initiator: 'https://example.com' },
    { url: 'https://api.example.com/data.json', type: 'xmlhttprequest', initiator: 'https://example.com' },
  ];

  const latencies: number[] = [];

  for (let i = 0; i < iterations; i++) {
    for (const req of testUrls) {
      const start = performance.now();
      wasm.should_block(req.url, req.type, req.initiator);
      const end = performance.now();
      latencies.push((end - start) * 1000);
    }
  }

  latencies.sort((a, b) => a - b);
  return latencies;
}

function percentile(sorted: number[], p: number): number {
  const idx = Math.ceil(sorted.length * p) - 1;
  return sorted[Math.max(0, idx)] ?? 0;
}

async function main(): Promise<void> {
  console.log('Performance Budget Check');
  console.log('='.repeat(50));
  console.log();

  const results: Budget[] = [];

  console.log('Loading WASM module...');
  const wasm = await loadWasm();

  console.log('Loading snapshot...');
  const snapshot = await loadSnapshot();

  const snapshotSizeMb = snapshot.byteLength / (1024 * 1024);
  results.push({
    name: 'Snapshot Size',
    limit: BUDGETS.snapshotSizeMb,
    unit: 'MB',
    actual: snapshotSizeMb,
    passed: snapshotSizeMb <= BUDGETS.snapshotSizeMb,
  });

  console.log('Measuring cold start...');
  const coldStartMs = measureColdStart(wasm, snapshot);
  results.push({
    name: 'Cold Start',
    limit: BUDGETS.coldStartMs,
    unit: 'ms',
    actual: coldStartMs,
    passed: coldStartMs <= BUDGETS.coldStartMs,
  });

  console.log('Warming up...');
  for (let i = 0; i < 1000; i++) {
    wasm.should_block('https://example.com/test', 'script', 'https://example.com');
  }

  console.log('Measuring match latency (10000 iterations)...');
  const latencies = measureMatchLatency(wasm, 2000);
  const p99Us = percentile(latencies, 0.99);
  results.push({
    name: 'Match P99 Latency',
    limit: BUDGETS.matchP99Us,
    unit: 'μs',
    actual: p99Us,
    passed: p99Us <= BUDGETS.matchP99Us,
  });

  const wasmInfo = wasm.get_snapshot_info();
  const wasmPeakMb = wasmInfo.size / (1024 * 1024);
  results.push({
    name: 'WASM Memory Peak',
    limit: BUDGETS.wasmPeakMb,
    unit: 'MB',
    actual: wasmPeakMb,
    passed: wasmPeakMb <= BUDGETS.wasmPeakMb,
  });

  console.log();
  console.log('Results');
  console.log('-'.repeat(50));

  let allPassed = true;
  for (const result of results) {
    const status = result.passed ? '✓' : '✗';
    const color = result.passed ? '\x1b[32m' : '\x1b[31m';
    const reset = '\x1b[0m';
    console.log(
      `${color}${status}${reset} ${result.name}: ${result.actual.toFixed(2)} ${result.unit} (limit: ${result.limit} ${result.unit})`
    );
    if (!result.passed) {
      allPassed = false;
    }
  }

  console.log();
  console.log('='.repeat(50));

  if (allPassed) {
    console.log('\x1b[32m✓ All performance budgets passed\x1b[0m');
    process.exit(0);
  } else {
    console.log('\x1b[31m✗ Performance budget exceeded\x1b[0m');
    process.exit(1);
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

File: crates/bb-cli/src/perf_budget.rs
use std::path::Path;
use std::time::Instant;

use bb_core::matcher::Matcher;
use bb_core::psl::get_etld1;
use bb_core::snapshot::Snapshot;
use bb_core::types::{MatchDecision, RequestContext, RequestType, SchemeMask};
use bb_core::url::{extract_host, extract_scheme};

use crate::snapshot;

pub struct PerfBudgetOptions {
    pub input_paths: Vec<String>,
    pub snapshot_path: String,
    pub compile: bool,
}

struct BudgetRequest {
    url: String,
    request_type: String,
    initiator: Option<String>,
}

const BUDGET_COLD_START_MS: f64 = 500.0;
const BUDGET_WASM_PEAK_MB: f64 = 50.0;
const BUDGET_MATCH_P99_US: f64 = 1000.0;
const BUDGET_SNAPSHOT_MB: f64 = 30.0;

pub fn run_perf_budget(opts: PerfBudgetOptions) -> Result<(), String> {
    println!("Performance Budget Check");
    println!("==================================================");

    let snapshot_path = Path::new(&opts.snapshot_path);
    let snapshot_bytes = if opts.compile {
        let (bytes, stats) = snapshot::compile_snapshot_bytes(&opts.input_paths, false)?;
        snapshot::write_snapshot(snapshot_path, &bytes)?;
        println!(
            "Compiled {} list(s): {} -> {} rules",
            opts.input_paths.len(),
            stats.rules_before,
            stats.rules_after
        );
        bytes
    } else {
        snapshot::read_snapshot(snapshot_path)?
    };

    let snapshot_size_mb = snapshot_bytes.len() as f64 / (1024.0 * 1024.0);

    println!("Loading snapshot...");
    let cold_start_begin = Instant::now();
    let snapshot = Snapshot::load(&snapshot_bytes)
        .map_err(|e| format!("Invalid snapshot: {}", e))?;
    let matcher = Matcher::new(&snapshot);
    let cold_start_ms = cold_start_begin.elapsed().as_secs_f64() * 1000.0;

    println!("Warming up...");
    let warm_req = BudgetRequest {
        url: "https://example.com/test".to_string(),
        request_type: "script".to_string(),
        initiator: Some("https://example.com".to_string()),
    };
    for _ in 0..1000 {
        let _ = should_block(&matcher, &warm_req);
    }

    println!("Measuring match latency...");
    let latencies = measure_match_latency(&matcher, 2000);
    let p99_us = percentile(&latencies, 0.99);

    let wasm_peak_mb = snapshot_size_mb;

    let mut passed = true;
    println!();
    println!("Results");
    println!("--------------------------------------------------");

    passed &= report_budget("Snapshot Size", snapshot_size_mb, BUDGET_SNAPSHOT_MB, "MB");
    passed &= report_budget("Cold Start", cold_start_ms, BUDGET_COLD_START_MS, "ms");
    passed &= report_budget("Match P99 Latency", p99_us, BUDGET_MATCH_P99_US, "μs");
    passed &= report_budget("WASM Memory Peak", wasm_peak_mb, BUDGET_WASM_PEAK_MB, "MB");

    println!();
    println!("==================================================");

    if passed {
        println!("✓ All performance budgets passed");
        Ok(())
    } else {
        Err("Performance budget exceeded".to_string())
    }
}

fn report_budget(name: &str, actual: f64, limit: f64, unit: &str) -> bool {
    let passed = actual <= limit;
    let status = if passed { "✓" } else { "✗" };
    println!(
        "{} {}: {:.2} {} (limit: {:.2} {})",
        status, name, actual, unit, limit, unit
    );
    passed
}

fn measure_match_latency(matcher: &Matcher, iterations: usize) -> Vec<f64> {
    let test_urls = [
        BudgetRequest {
            url: "https://pagead2.googlesyndication.com/pagead/js/adsbygoogle.js".to_string(),
            request_type: "script".to_string(),
            initiator: Some("https://example.com".to_string()),
        },
        BudgetRequest {
            url: "https://www.google-analytics.com/analytics.js".to_string(),
            request_type: "script".to_string(),
            initiator: Some("https://example.com".to_string()),
        },
        BudgetRequest {
            url: "https://example.com/style.css".to_string(),
            request_type: "stylesheet".to_string(),
            initiator: Some("https://example.com".to_string()),
        },
        BudgetRequest {
            url: "https://cdn.example.com/image.png".to_string(),
            request_type: "image".to_string(),
            initiator: Some("https://example.com".to_string()),
        },
        BudgetRequest {
            url: "https://api.example.com/data.json".to_string(),
            request_type: "xmlhttprequest".to_string(),
            initiator: Some("https://example.com".to_string()),
        },
    ];

    let mut latencies = Vec::new();

    for _ in 0..iterations {
        for req in &test_urls {
            let start = Instant::now();
            let _ = should_block(matcher, req);
            let elapsed = start.elapsed().as_secs_f64() * 1_000_000.0;
            latencies.push(elapsed);
        }
    }

    latencies.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    latencies
}

fn percentile(sorted: &[f64], p: f64) -> f64 {
    if sorted.is_empty() {
        return 0.0;
    }
    let idx = ((sorted.len() as f64) * p).ceil() as usize;
    let idx = idx.saturating_sub(1).min(sorted.len() - 1);
    sorted[idx]
}

fn should_block(matcher: &Matcher, req: &BudgetRequest) -> bool {
    match_request(matcher, req).decision == MatchDecision::Block
}

fn match_request(matcher: &Matcher, req: &BudgetRequest) -> bb_core::types::MatchResult {
    let req_host = extract_host(&req.url).unwrap_or("");
    let req_etld1 = get_etld1(req_host);

    let is_main_frame = req.request_type == "main_frame" || req.request_type == "document";
    let site_url = if is_main_frame {
        req.url.as_str()
    } else {
        req.initiator.as_deref().unwrap_or(req.url.as_str())
    };
    let site_host = extract_host(site_url).unwrap_or(req_host);
    let site_etld1 = get_etld1(site_host);

    let scheme = extract_scheme(&req.url).unwrap_or(SchemeMask::HTTP);
    let is_third_party = !site_etld1.is_empty() && req_etld1 != site_etld1;
    let request_type = RequestType::from_str(&req.request_type);

    let ctx = RequestContext {
        url: &req.url,
        req_host,
        req_etld1: &req_etld1,
        site_host,
        site_etld1: &site_etld1,
        is_third_party,
        request_type,
        scheme,
        tab_id: 1,
        frame_id: 0,
        request_id: "perf",
    };

    matcher.match_request(&ctx)
}
